saml/saml.php

```diff
@@ -0,0 +1,429 @@
+        DI::config()->get('saml', 'sp_key') &&
+        DI::config()->get('saml', 'sp_cert') &&
+        DI::config()->get('saml', 'idp_cert');
+
+}
+
+function saml_sso_initiate(&$a, &$b) {
+    if (!saml_is_configured()) return;
+
+    $auth = new \OneLogin\Saml2\Auth(saml_settings());
+    $ssoBuiltUrl = $auth->login(null, array(), false, false, true);
+    $_SESSION['AuthNRequestID'] = $auth->getLastRequestID();
+    header('Pragma: no-cache');
+    header('Cache-Control: no-cache, must-revalidate');
+    header('Location: ' . $ssoBuiltUrl);
+    exit();
+}
+
+function saml_sso_reply($a) {
+    $auth = new \OneLogin\Saml2\Auth(saml_settings());
+    $requestID = null;
+
+    if (isset($_SESSION) && isset($_SESSION['AuthNRequestID'])) {
+        $requestID = $_SESSION['AuthNRequestID'];
+    }
+
+    $auth->processResponse($requestID);
+    unset($_SESSION['AuthNRequestID']);
+
+    $errors = $auth->getErrors();
+
+    if (!empty($errors)) {
+	echo "Errors encountered.";
+        Logger::error(implode(', ', $errors));
+        exit();
+    }
+
+    if (!$auth->isAuthenticated()) {
+        echo "Not authenticated";
+        exit();
+    }
+
+    $username = $auth->getNameId();
+    $email = $auth->getAttributeWithFriendlyName('email')[0];
+    $name = $auth->getAttributeWithFriendlyName('givenName')[0];
+    $last_name = $auth->getAttributeWithFriendlyName('surname')[0];
+
+    if (strlen($last_name)) {
+        $name .= " $last_name";
+    }
+
+    if (!DBA::exists('user', ['nickname' => $username])) {
+        $user = saml_create_user($username, $email, $name);
+    } else {
+        $user = User::getByNickname($username);
+    }
+
+    if (!empty($user['uid'])) {
+        DI::auth()->setForUser($a, $user);
+    }
+
+    if (isset($_POST['RelayState'])
+        && \OneLogin\Saml2\Utils::getSelfURL() != $_POST['RelayState'])
+    {
+        $auth->redirectTo($_POST['RelayState']);
+    }
+}
+
+function saml_slo_initiate(&$a, &$b) {
+    $auth = new \OneLogin\Saml2\Auth(saml_settings());
+
+    $sloBuiltUrl = $auth->logout();
+    $_SESSION['LogoutRequestID'] = $auth->getLastRequestID();
+    header('Pragma: no-cache');
+    header('Cache-Control: no-cache, must-revalidate');
+    header('Location: ' . $sloBuiltUrl);
+    exit();
+}
+
+function saml_slo_reply() {
+    $auth = new \OneLogin\Saml2\Auth(saml_settings());
+
+    if (isset($_SESSION) && isset($_SESSION['LogoutRequestID'])) {
+        $requestID = $_SESSION['LogoutRequestID'];
+    } else {
+        $requestID = null;
+    }
+
+    $auth->processSLO(false, $requestID);
+
+    $errors = $auth->getErrors();
+
+    if (empty($errors)) {
+        $auth->redirectTo(DI::baseUrl());
+    } else {
+        Logger::error(implode(', ', $errors));
+    }
+}
+
+function saml_input($key, $label, $description) {
+    return [
+        '$' . $key => [
+            $key,
+            DI::l10n()->t($label),
+            DI::config()->get('saml', $key),
+            DI::l10n()->t($description),
```

Comment:
Unfortunately, these dynamic translation strings won't be picked up by our translation string collector script and as a result won't be available for our volunteer translators to translate in Transifex. You have to explicitly wrap each translation string with a `DI::l10n()->t()` call.